### PR TITLE
[GitHub Repository Search] Add dropdown for filtering by user/org

### PR DESCRIPTION
**Pull request by jberglinds**
## Description
Adds a dropdown to the search field for filtering the search by user/orgs, similarly to how the native "Search Repositories" extension works. Also, includes forks when searching for repositories like the native extension.

* "All Repositories" -> The same as before, searching all of Github (Default selection)
* "My Repositories" -> Searches for personal repos and repos of organizations that you are a member of
* "{Github username}" -> Searches only for personal repos
* "{Organization username}" -> Searches only for repos of the chosen org

## Screencast
github-repository-search 2022-07-14 at 16 05 04
github-repository-search 2022-07-14 at 16 05 56
github-repository-search 2022-07-14 at 16 03 31

## Checklist

- [ ] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [ ] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [ ] I checked that files in the `assets` folder are used by the extension itself
- [ ] I checked that assets used by the `README` are placed outside of the `metadata` folder

### CODE REVIEW

File: extensions/github-repository-search/src/index.tsx [hunk 1]
```diff
@@ -95,7 +153,7 @@ function Actions(props: { repository: Repository; onVisit: (repository: Reposito
           icon={{ source: "pull-request.png", tintColor: Color.PrimaryText }}
           title="Open Pull Requests"
           url={`${props.repository.url}/pulls`}
-          shortcut={{ modifiers: ["cmd"], key: "p" }}
+          shortcut={{ modifiers: ["cmd", "shift"], key: "p" }}
```
Comment (jberglinds) [inline extensions/github-repository-search/src/index.tsx:156]: cmd+p is the default hotkey to trigger the dropdown so this had to change

File: extensions/github-repository-search/src/index.tsx [hunk 2]
```diff
@@ -121,7 +179,7 @@ function Actions(props: { repository: Repository; onVisit: (repository: Reposito
             icon={{ source: "project.png", tintColor: Color.PrimaryText }}
             title="Open Projects"
             url={`${props.repository.url}/projects`}
-            shortcut={{ modifiers: ["cmd", "shift"], key: "p" }}
+            shortcut={{ modifiers: ["cmd", "shift", "opt"], key: "p" }}
```
Comment (jberglinds) [inline extensions/github-repository-search/src/index.tsx:182]: this is quite a lot, open to suggestions

File: extensions/github-repository-search/src/index.tsx [hunk 3]
```diff
@@ -11,55 +11,113 @@ import {
...
         subtitle={visitedRepositories ? String(visitedRepositories.length) : undefined}
       >
         {visitedRepositories
-          ?.filter((r) => r.nameWithOwner.includes(searchText ?? ""))
+          ?.filter(
+            (r) =>
+              r.nameWithOwner.includes(searchText ?? "") &&
+              r.nameWithOwner.match(
+                // Converting query filter string to regexp:
+                `${searchFilter?.replaceAll(/org:|user:/g, "").replaceAll(" ", "|")}\/.*`
```
Comment (jberglinds) [inline extensions/github-repository-search/src/index.tsx:74]: not proud of this but the alternative is quite some extra state to manage

**Comment by DonDebonair** [comment]
This looks great! How can we help test this?

**Comment by jberglinds** [comment]
As this week is "launch week" at raycast I'm guessing that Thomas is really busy. 

If you want to test it out before it gets merged you can clone the changes and run `npm install && npm build` and it should install as a separate extension, @DonDebonair

**Comment by DonDebonair** [comment]
I tested it (need to run `npm run dev` to get it to work btw) and it works great! The added option to search only your own repos or even within a specific org, together with caching visited repos, is exactly what I need!

Would be nice if this could be merged and published soon :)

**Comment by thomaspaulmann** [comment]
Sorry that this drags so long. Been busy with Launch Week. Gonna pick it up next week.

**Comment by jberglinds** [comment]
Hey @thomaspaulmann, any chance you could take a look at this soon-ish?

File: extensions/github-repository-search/src/index.tsx (hunk 1 above)
Comment (thomaspaulmann) [reply extensions/github-repository-search/src/index.tsx:156]: Makes sense, thanks!

File: extensions/github-repository-search/tsconfig.json [hunk 4]
```diff
@@ -3,7 +3,7 @@
   "display": "Node 16",
   "include": ["src/**/*"],
   "compilerOptions": {
-    "lib": ["es2020"],
+    "lib": ["es2021"],
```
Comment (thomaspaulmann) [inline extensions/github-repository-search/tsconfig.json:6]: Why did you change this?

File: extensions/github-repository-search/src/useRepositories.ts [hunk 5]
```diff
@@ -53,7 +53,9 @@ export function useRepositories(searchText: string | undefined) {
       setState((oldState) => ({ ...oldState, isLoading: true }));
 
       try {
-        const { search } = await octokit.graphql<SearchRepositoriesResponse>(SEARCH_REPOSITORIES_QUERY, { searchText });
+        const { search } = await octokit.graphql<SearchRepositoriesResponse>(SEARCH_REPOSITORIES_QUERY, {
+          searchQuery: `${searchFilter} fork:true ${searchText}`,
```
Comment (thomaspaulmann) [inline extensions/github-repository-search/src/useRepositories.ts:57]: Should it be a preference if you want to include forks in the search? It's a bit confusing that you have `searchText` and `searchFilter`. Wouldn't it be easier to inject the entire `searchQuery` from outside the hook?

File: extensions/github-repository-search/src/index.tsx (hunk 2 above)
Comment (thomaspaulmann) [reply extensions/github-repository-search/src/index.tsx:182]: I guess it's fine.

File: extensions/github-repository-search/src/index.tsx [hunk 6]
```diff
@@ -11,55 +11,113 @@ import {
...
       </List.Section>
     </List>
   );
 }
 
+function FilterDropdown({
+  userData,
+  onFilterChange,
+}: {
+  userData: UserDataResponse["viewer"] | undefined;
```
Comment (thomaspaulmann) [inline extensions/github-repository-search/src/index.tsx:97]: Isn't that the `viewer`?
```suggestion
  viewer?: UserDataResponse["viewer"];
```

File: extensions/github-repository-search/src/index.tsx [hunk 7]
```diff
@@ -11,55 +11,113 @@ import {
...
+        <List.Dropdown.Section>
+          <List.Dropdown.Item title={userData.login} value={`user:${userData.login}`} />
+          {userData.organizations.nodes.map((org) => (
+            <List.Dropdown.Item key={org.login} title={org.login} value={`org:${org.login}`} />
+          ))}
+        </List.Dropdown.Section>
+      )}
+    </List.Dropdown>
+  );
+}
```
Comment (thomaspaulmann) [inline extensions/github-repository-search/src/index.tsx:119]: How about not returning a dropdown if we don't have any user data? A dropdown with just a single item doesn't make much sense.
```suggestion
  return userData ? (
    <List.Dropdown tooltip="Filter Repositories" storeValue={true} onChange={onFilterChange}>
      <List.Dropdown.Item title="All Repositories" value="" />
        <List.Dropdown.Item
          title="My Repositories"
          value={`user:${userData.login} ${userData.organizations.nodes.map((org) => `org:${org.login}`).join(" ")}`}
        />
        <List.Dropdown.Section>
          <List.Dropdown.Item title={userData.login} value={`user:${userData.login}`} />
          {userData.organizations.nodes.map((org) => (
            <List.Dropdown.Item key={org.login} title={org.login} value={`org:${org.login}`} />
          ))}
        </List.Dropdown.Section>
    </List.Dropdown> : null
  );
}
```

File: extensions/github-repository-search/src/index.tsx [hunk 8]
```diff
@@ -11,55 +11,113 @@ import {
...
       >
         {visitedRepositories
-          ?.filter((r) => r.nameWithOwner.includes(searchText ?? ""))
+          ?.filter(
+            (r) =>
+              r.nameWithOwner.includes(searchText ?? "") &&
+              r.nameWithOwner.match(
+                // Converting query filter string to regexp:
+                `${searchFilter?.replaceAll(/org:|user:/g, "").replaceAll(" ", "|")}/.*`
+              )
```
Comment (thomaspaulmann) [inline extensions/github-repository-search/src/index.tsx:75]: Shouldn't that be a disjunction?
```suggestion
              r.nameWithOwner.includes(searchText ?? "") ||
              r.nameWithOwner.match(
                // Converting query filter string to regexp:
                `${searchFilter?.replaceAll(/org:|user:/g, "").replaceAll(" ", "|")}/.*`
              )
```

**Review by thomaspaulmann — CHANGES_REQUESTED** [review]
I left a few comments in the PR. My main concern is the new required scopes. People that don't have a PAT with the user and org scope will constantly get failure toasts. Given that the extension is used by a lot of people, we should find a way to work around this.

github-repository-search 2022-07-29 at 10 37 46

A few ideas:
- Somehow check the scope and do the right thing. If user/org scope is available, add your changes, if not, then stick with the previous behavior.
- Put your changes behind a new preference that people have to opt-in. I don't like preferences for such things.

We have this problem with other extensions and need some elegant way to solve it. I'm open for suggestions here.

**Comment by thomaspaulmann** [comment]
@jberglinds left a few comments. Sorry that it took me longer than expected.

File: extensions/github-repository-search/tsconfig.json (hunk 4 above)
Comment (jberglinds) [reply extensions/github-repository-search/tsconfig.json:6]:
```
src/index.tsx:74:34 - error TS2550: Property 'replaceAll' does not exist on type 'string'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2021' or later.
```

File: extensions/github-repository-search/src/index.tsx (hunk 8 above)
Comment (jberglinds) [reply extensions/github-repository-search/src/index.tsx:75]: Why?
The second filter just targets the owner part of the repository id. Even if you get a hit with the first filter on something I think it should still filter the results by whatever you have selected in the dropdown

File: extensions/github-repository-search/src/index.tsx (hunk 6 above)
Comment (jberglinds) [reply extensions/github-repository-search/src/index.tsx:97]: Sure it is.
I renamed it because I didn't think 'viewer' was a great name anymore when you detach it from the context of the request, but I guess you still have it in the type information

File: extensions/github-repository-search/src/useRepositories.ts (hunk 5 above)
Comment (jberglinds) [reply extensions/github-repository-search/src/useRepositories.ts:57]: Sure, can do it

**Comment by jberglinds** [comment]
I believed I've addressed most of your comments. 
Regarding the error messages I'm not sure what to do about it. I think the error message it pretty clear about whats wrong.

**Comment by thomaspaulmann** [comment]
It isn't about that the error message isn't clear. It's about that existing users see this message from one day to another. This causes confusion.

I have an idea to be smarter:
- Show a one-time view to explain the new features and mention that you need a token with the expected scopes
- Build the API requests in a more robust way
  - Essentially, if the user doesn't have user scope, don't search private repos. And if the user's scope doesn't have access to orgs, don't search them.
  - The dropdown would then be optional.

Benefit is that existing users can keep using the extension as is and don't need to change the scope of the token. Happy to help implementing this. WDYT?

**Comment by jberglinds** [comment]
I've had absolutely no time lately to look at this. Please help out if you can find the time 🙏

**Comment by thomaspaulmann** [comment]
I started to work on the changes in [this branch](https://github.com/raycast/extensions/tree/pr/2266). For now I upgraded to the latest APIs. Next, I'll convert the history to use our utils and then I'll look into the graceful handling of the tokens.

**Comment by thomaspaulmann** [comment]
@jberglinds I found the time to upgrade the extension. I created a new PR for it, it was easier after all the additional changes. I'm closing this PR for #2901. Please give it a try and let me know if there is anything else that you would need.
